groups/reconcile.go

```diff
@@ -0,0 +1,397 @@
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	// timeout the request after 30 seconds
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	resp, err := client.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	return ioutil.ReadAll(resp.Body)
+}
+
+func printGroupMembersAndSettings(srv *admin.Service, srv2 *groupssettings.Service) error {
+	g, err := srv.Groups.List().Customer("my_customer").OrderBy("email").Do()
+	if err != nil {
+		return fmt.Errorf("unable to retrieve users in domain: %v", err)
+	}
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
+	for _, g := range g.Groups {
+		// Don't touch existing mailing lists, we should
+		// always prefix with "k8s-infra-"
+		if !strings.HasPrefix(g.Email, "k8s-infra-") {
+			continue
+		}
+		log.Printf("%s\n", g.Email)
+
+		g2, err := srv2.Groups.Get(g.Email).Do()
+		if err != nil {
+			return fmt.Errorf("unable to retrieve group info for group %s: %v", g.Email, err)
+		}
+		log.Printf(">> Allow external members %s\n", g2.AllowExternalMembers)
+
+		l, err := srv.Members.List(g.Email).Do()
+		if err != nil {
+			return fmt.Errorf("unable to retrieve members in group : %v", err)
+		}
+
+		if len(l.Members) == 0 {
+			log.Println("No members found in group.")
+		} else {
+			for _, m := range l.Members {
+				log.Printf(">>> %s (%s)\n", m.Email, m.Role)
+			}
+		}
+		log.Printf("\n")
+
+	}
+	return nil
+}
+
+func createGroupIfNecessary(srv *admin.Service, groupEmailId string, description string) error {
+	_, err := srv.Groups.Get(groupEmailId).Do()
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			if config.DryRun {
+				log.Printf("dry-run : skipping creation of group %s\n", groupEmailId)
+			} else {
+				log.Printf("Trying to create group: %s\n", groupEmailId)
+				g4, err := srv.Groups.Insert(&admin.Group{
+					Email:       groupEmailId,
+					Name:        description,
+					Description: "Kubernetes wg-k8s-infra test group #2",
+				}).Do()
+				if err != nil {
+					return fmt.Errorf("unable to add new group %s: %v", groupEmailId, err)
+				}
+				log.Printf("> Successfully created group %s\n", g4.Email)
+			}
+		} else {
+			return fmt.Errorf("unable to fetch group: %#v", err.Error())
+		}
+	}
+	return nil
+}
+
+func deleteGroupsIfNecessary(service *admin.Service) error {
+	g, err := service.Groups.List().Customer("my_customer").OrderBy("email").Do()
+	if err != nil {
+		return fmt.Errorf("unable to retrieve users in domain: %v", err)
+	}
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
+	for _, g := range g.Groups {
+		// Don't touch existing mailing lists, we should
+		// always prefix with "k8s-infra-"
+		if !strings.HasPrefix(g.Email, "k8s-infra-") {
```

Comment:
I was freaked out we'd be nuking everything until I read this far, maybe pull into function's interface to make it clear there is a blacklist

Is there a way to constrain this to groups created solely by this service account instead of or in addition to blacklisting by name?

Reply:
So, eventually, PST would like to use something like this for automating memberships to things like the distributors-announce list.

I think eventually, I'd suggest going with either a complete authoritative model (all mailing lists through k8s.io are managed this way, similar to peribolos), or an opportunistic model like label_sync (we create and manage groups that are defined, ignore the ones we don't, and only delete if we have a `deleteAfter` statement defined saying we should explicitly remove a group. Whitelisting/blacklisting based on strings isn't ideal IMO.

That said, not blocking for me from an MVP.

Reply:
Added to follow up issue https://github.com/kubernetes/k8s.io/issues/255